File: src/main/java/common/optaplanner/basedomain/Coordinates.java
package common.optaplanner.basedomain;

public class Coordinates implements BaseCoordinate {

    private double latitude;
    private double longitude;

    public Coordinates(double latitude, double longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    @Override
    public Double getLatitude() {
        return latitude;
    }

    @Override
    public Double getLongitude() {
        return longitude;
    }

    @Override
    public Double latitude() {
        return getLatitude();
    }

    @Override
    public Double longitude() {
        return getLongitude();
    }
}

File: src/main/java/common/basedomain/coordinates.kt
package common.basedomain

data class Coordinates(val latitude:Double, val longitude: Double): BaseCoordinate {
    override fun latitude():Double {
        return latitude //To change body of created functions use File | Settings | File Templates.
    }

    override fun longitude():Double {
        return longitude //To change body of created functions use File | Settings | File Templates.
    }
}

data class XyCoordinates(val x: Double, val y: Double): BaseCoordinate {
    override fun latitude():Double {
        return x
    }

    override fun longitude():Double{
        return y
    }
}

interface BaseCoordinate {
    fun latitude(): Double
    fun longitude(): Double
}